### ISSUE TITLE
Bottom-up confmap/PAF loss (plain MSELoss) has no fg/bg weighting — collapses on large images with sparse targets

### ISSUE THREAD

**Issue by eberrigan** [talmolab/sleap-nn#718]
## Summary

`BottomUpLightningModule`'s confmap + PAF loss is a plain, unweighted `nn.MSELoss()` with no foreground/background weighting. On large images with sparse keypoint targets, this causes training to collapse toward predicting near-zero everywhere (no detections), which we hit concretely on a real dataset — forcing a downscale + coarse-`output_stride` workaround that costs a lot of localization precision. Filing so this becomes a tracked, fixable gap rather than a silent per-project workaround.

## Concrete evidence (talmolab/sleap-roots-training#33)

Dataset: Arabidopsis primary-root, multi-plant cylinder, 1088×2048 images, 6-node skeleton (`r1..r6`), bottom-up (confmaps + PAFs). `sleap-nn` pinned `>=0.2.0,<0.3.0`.

**TF reference** (legacy `sleap-train`, committed fixtures in `talmolab/sleap-roots-training`'s `tests/fixtures/tf_reference/`) trained at **full resolution** (`input_scaling: 1`, no downscaling) with confmap **`output_stride` held constant at 2** and PAF `output_stride` at 8 (sigma 2.5 / 75 respectively — identical head hyperparameters to what we used below), across a sweep of `model.backbone.unet.max_stride` ∈ {8, 16, 32, 64} (backbone depth/receptive field — a separate, independent parameter from `output_stride`, which was never varied). All runs with usable summaries achieved `dist_avg` **0.99–2.08 px**.

**sleap-nn 0.2.0 bottom-up at `output_stride: 2`** (same confmap resolution as every TF run above, but with the input additionally downscaled 0.5× — i.e. a *lower* effective resolution than any TF run used) **collapsed to zero predicted instances (all-NaN) on 2 of 3 same-config seeds**; the one seed that did produce output missed ~43% of ground-truth instances.

Escaping the collapse required both downscaling the input (`scale: 0.5`) *and* moving to the coarser `output_stride: 4` (finer than TF never needed to go) + online hard-keypoint mining. That "works" (all instances detected, stable across 3 seeds) but lands at `dist_avg` **30.1–37.8 px** — **~15–38× worse** than the TF reference on the *same dataset*, which trained successfully at the finer `output_stride: 2` without ever needing to touch input resolution.

Full write-up with per-seed numbers: [`docs/training.md`, "PyTorch baseline" section](https://github.com/talmolab/sleap-roots-training/blob/main/docs/training.md) and the "Finding: `output_stride 2` collapses" subsection.

## Root cause (read from source)

`sleap_nn/training/lightning_modules.py`, `BottomUpLightningModule.training_step`/`validation_step`:

```python
confmap_loss = nn.MSELoss()(confmaps, y_confmap)
pafs_loss = nn.MSELoss()(pafs, y_paf)
```

No foreground weighting, no focal-style loss. On a 1088×2048 image, Gaussian keypoint confmap targets are a tiny fraction of the frame (well under 1% foreground), so a flat per-pixel MSE is dominated by trivially correct background, and the training signal for actual peaks is weak — consistent with the fg/bg imbalance #698 diagnosed elsewhere.

## Related: #698

#698 ("log confmap fg/bg MSE split as a training diagnostic", merged 2026-07-13) independently confirmed this exact imbalance on the MARS mouse-pose dataset (~1600–1800× fg/bg MSE ratio) and added `{train,val}/confmap_loss_fg` / `confmap_loss_bg` / `confmap_fg_frac` logging. However, per that PR's own description:
- It's **diagnostic only** — computed under `torch.no_grad()`, not part of the optimized loss.
- It's wired into `SingleInstanceLightningModule`, `TopDownCenteredInstanceLightningModule`, and `CentroidLightningModule` only — **explicitly not `BottomUpLightningModule`**.
- Its own "Future Considerations" section flags a weighted-loss or focal-style loss as a follow-up, "not in scope here."

That follow-up doesn't appear to exist yet (searched issues/PRs for `bottomup`/`focal`/`weighted loss` — nothing found beyond #698 itself).

## Ask

1. Extend the #698 fg/bg diagnostic (`confmap_loss_fg`/`confmap_loss_bg`/`confmap_fg_frac`) to `BottomUpLightningModule` (and ideally `BottomUpMultiClassLightningModule`) so this imbalance is visible for bottom-up training too, not just the three heads already covered.
2. Consider an opt-in weighted or focal-style loss for confmap/PAF heads (e.g. a foreground/background weight ratio, or focal loss) so bottom-up models can train at finer `output_stride` / full input resolution on large images without the downscale-to-escape-collapse workaround — since that workaround directly trades away localization precision that TF didn't need to sacrifice on this same dataset.

cc @anirudh9280, who hit this while establishing the Tier 1 PyTorch baseline in talmolab/sleap-roots-training#33.

**Comment by eberrigan** [comment]
Correction to the original post: the TF sweep varied `model.backbone.unet.max_stride` (8/16/32/64 — backbone depth), not confmap `output_stride`, which was held constant at 2 across all TF runs. Updated the issue body above to reflect this — it actually sharpens the comparison (TF trained successfully at output_stride 2 / full resolution; sleap-nn collapsed at that same output_stride even with the input already downscaled).

**Comment by anirudh9280** [comment]
Adding a reproduction that isolates this to the loss. I translated the legacy TF config for this dataset (the tests/fixtures/tf_reference settings) 1:1 into sleap-nn 0.2.0 and ran it at full resolution, so nothing on our side differs from the TF run that trained to roughly 1 to 2 px except the backend. Settings: scale 1.0, max_stride 16, filters 24 / filters_rate 1.5, output_stride 2, confmaps.sigma 2.5 / pafs.sigma 75, online_mining false, Adam lr 1e-4.

Result on all 3 seeds (W&B 7ejsj2x6 / cy1rkq73 / d5l6qvmf): each trains to convergence (no OOM, 727K-param model) but predicts 0 instances on both train and val, all-NaN metrics, with val/confmaps_loss ~7e-05. Full resolution plus TF's own architecture and sigmas rules out a resolution ceiling and a bad config on our end, so the remaining difference from the working TF run is the bottom-up loss.

This is consistent with a controlled sigma ablation on the same data (same config, only confmaps.sigma changed): at output_stride 2, sigma 2.5 collapses (near-zero val/confmaps_loss) but sigma 5.0 trains stably (val/confmaps_loss ~0.0047). That sensitivity is what you would expect from an unweighted per-pixel MSE over a tiny foreground fraction. A foreground-weighted or focal-style confmap loss would make bottom-up robust to sigma / output_stride and let it train at higher resolution without collapsing.

Full config and per-seed numbers are in talmolab/sleap-roots-training#33.

**Comment by anirudh9280** [comment]
Correction / update on the reproduction I posted above, two things change the read.

1. As you noted, neither current sleap-nn nor legacy TF SLEAP has foreground-weighted confmap/PAF loss. Since both lack it and only sleap-nn collapses, no foreground weighting cannot by itself be what differentiates the backends.

2. Our TF-parity control was confounded: it matched TF on architecture, resolution, and sigma but kept sleap-nn's short training schedule (min_train_steps ~25) instead of TF's ~8x-oversampled schedule (min_train_steps_per_epoch 200, ~2000-step early-stop patience). So TF may simply have had far more gradient steps to escape the predict-background minimum before early stopping, with no loss-formula difference required.

We are running a schedule-matched control now (TF architecture / full res / sigma 2.5 / output_stride 2 AND min_train_steps_per_epoch 200 / matching patience). If it still collapses with everything matched, that is real evidence for a loss or training-dynamics cause. If it trains stably, the earlier collapse was a step-count effect and this issue's framing should change. I will report the result here. Until then, please treat the unweighted-loss-causes-collapse claim as unresolved rather than established.

**Comment by anirudh9280** [comment]
Result from the schedule-matched control, all 3 seeds, and it points away from the loss formula.

At the settings that collapsed on all 3 seeds before (full res, output_stride 2, sigma 2.5, TF architecture), changing only min_train_steps_per_epoch from ~25 to 200 (matching TF's schedule and early-stop patience) makes all 3 train: train dist_avg 6.3 to 9.7 px, p50 2.6 to 4.2 px, no collapse on any seed. Since both sleap-nn and legacy TF use unweighted MSE, and the collapse disappears once the step budget matches TF, the unweighted loss is not what causes this collapse. It looks like sleap-nn's default min_train_steps_per_epoch / early-stopping is just too short for the sparse large-image regime, so training early-stops before the model climbs out of the all-background basin.

Also useful: val/confmaps_loss is not a reliable collapse indicator on these sparse targets. The healthy runs above sit at ~6e-5, the same near-zero range as the collapsed runs, because both an all-background and a sharp-peak prediction have tiny MSE. Matched-instance count is the reliable signal.

A foreground-weighted loss would still likely help convergence and robustness, so this is a reasonable enhancement, but it is probably not the fix for the collapse we reported. Happy to share the schedule-matched config or the runs.

**Comment by anirudh9280** [comment]
## Closing the loop on the reproduction I posted here, with one attribution fix

Following up on the schedule-matched control above. Two things to add, one of which corrects how I described the cause.

**The attribution was still wrong in my last comment.** I framed the collapse as sleap-nn's default schedule being too short relative to TF's. It is not. sleap-nn's default `min_train_steps_per_epoch` is 200 (`sleap_nn/config/trainer_config.py`, and `train_steps_per_epoch` then resolves to the max of that value and the data size). Our config explicitly set it to 25, which is 8x below your default. So the collapse I reported here came from our under-setting, and running sleap-nn at its own default would have avoided it entirely. The empirical result is unchanged, 25 collapses and 200 trains on all 3 seeds, but nothing about it is a sleap-nn defect. Credit to @eberrigan for catching both this and the original schedule confound.

**Independent confirmation via your own converter.** To take our hand-written config out of the loop, I ran `TrainingJobConfig.load_sleap_config()` on the legacy TF config for this dataset and trained the emitted native config directly. It defaults `min_train_steps_per_epoch` to 200 and carries TF's augmentation. All 3 seeds train, no collapse. So the "convert the legacy config" path does the right thing here with no intervention, which is worth knowing for anyone else migrating a TF bottom-up model.

### What may still be worth something to you

Even with the schedule fixed, the full-resolution / `output_stride 2` / `sigma 2.5` regime trains but under-detects on this dataset (Arabidopsis primary root, multi-plant cylinder, bottom-up, 6 nodes, 1088x2048 grayscale, 99 train and 21 val frames, 44 val instances):

| config | seed | instances detected | `dist_avg` | `dist_p50` | `vis_recall` |
|---|---|---|---|---|---|
| aug off, 200 steps | 42 | 38 / 44 | 30.77 | 7.77 | 0.632 |
| aug off, 200 steps | 43 | 34 / 44 | 16.21 | 5.67 | 0.382 |
| aug off, 200 steps | 44 | 32 / 44 | 14.91 | 6.47 | 0.380 |
| converter-derived, aug on, 200 steps | 42 | 42 / 44 | 33.55 | 7.84 | 0.647 |
| converter-derived, aug on, 200 steps | 43 | 38 / 44 | 24.25 | 7.21 | 0.601 |
| converter-derived, aug on, 200 steps | 44 | 28 / 44 | 21.83 | 12.17 | 0.423 |

Coarsening the targets, either `sigma 5.0` at `output_stride 2` or `scale 0.5` with `output_stride 4`, recovers detection to 44 of 44 with keypoint recall 0.85 to 0.95 on every seed. That pattern, where sparse tight targets train to convergence but under-detect while coarser targets do not, is at least consistent with a weak foreground gradient, so foreground weighting may still be a real improvement for large images with sparse keypoints. Some of that gap is also small-data overfitting on 99 frames, so I would not read the table as clean evidence for the loss either.

Either way it is an enhancement rather than the fix for what I originally reported here, and we are not blocked on it. Happy for this to be closed or re-scoped as an enhancement, whichever you prefer. The separate reporting problem in #719, where a 0-match eval produces all-NaN metrics and a warning cascade rather than a clear signal, is the one that actually cost us time and still stands on its own.

**Comment by anirudh9280** [comment]
Last follow up from our side, and it closes out the remaining thread here.

In my previous comment I said the full resolution / output_stride 2 / sigma 2.5 regime still under detects even with the schedule fixed, and floated that as possibly consistent with a weak foreground gradient. I then went and compared our numbers against the legacy TF reference properly, and two things came out of it that are worth recording here.

First, sleap-nn's evaluator is numerically identical to legacy SLEAP's. I took a legacy TF run's own saved labels_gt.val.slp and labels_pr.val.slp and scored them with sleap_nn.evaluation.Evaluator, then compared against the metrics.val.npz that SLEAP itself wrote for that run. All 11 metrics agree to every printed digit, dist avg 39.0026 and p50 8.4180 and p90 120.1750 and p95 162.9693 and p99 245.2260, vis recall 0.5583 with tp 134 and fn 106, mOKS 0.1906, oks_voc mAP 0.0111 and mAR 0.0250, and the same dists array of shape (40, 6) with 134 valid cells. The inference parameters match too, peak_threshold 0.2 on both sides, integral refinement at patch size 5, and identical PAF grouping constants (max_edge_length_ratio 0.25, dist_penalty_weight 1.0, n_points 10, min_line_scores 0.25, min_instance_peaks 0). So there is no evaluation or inference difference between the backends to explain anything.

Second, the apparent accuracy gap that motivated my speculation was a unit error on our side. The TF numbers we had been comparing against are millimeters, produced by a lab post processing step rather than by SLEAP. In pixels the TF sweep runs 21.65 to 39.00 dist_avg at vis_recall 0.495 to 0.872, which is the same regime our sleap-nn models are in, and our best config actually leads on detection at 0.89 to 0.95 recall with all 44 of 44 instances found.

There is also a data point here that cuts directly against the original framing of this issue. The stride 8 run in that legacy TensorFlow sweep collapsed, with 0 predicted instances on train, val and test. That is the same failure mode we reported here, in TensorFlow, with the same unweighted MSE that both backends share. Collapse on large sparse target images is not something sleap-nn does and legacy SLEAP does not.

Net for this issue. Nothing we have found is a sleap-nn defect. The collapse we originally reported was our own min_train_steps_per_epoch 25 against your default of 200, the evaluators agree exactly, and the inference defaults match. Foreground weighted loss may still be a genuine improvement for large images with sparse keypoints, but on our evidence it is an enhancement rather than a fix for anything reported here, and we are not blocked on it. Happy for this to be closed or re scoped as an enhancement, whichever you prefer. Thanks to @gitttt-1234 for the pointer to check the inference parameters, that is what sent us looking in the right place.
